Prompt beforeUnload (#4272)

gpltaylor opened the pull request:
Implement beforeUnload, allowing custom and standard alerts if a user tries to leave the site.
Includes docs and example.

Issue: (#4272)

gpltaylor commented:
Hi @mjackson here is the pull request for the Prompt.

This is tested on all Browsers/platforms using BrowserStack.
The only issue with this feature, if someone already has a custom beforeUnload function registered, this would replace their function. However, in this scenario, they would not use the beforeUnload prop.

For the sake of KISS/YAGNI I have not cached any existing registered events.

Let me know what you think.

mjackson commented on packages/react-router-website/api/Prompt.md:
````diff
@@ -34,3 +34,24 @@ Instead of conditionally rendering a `<Prompt>` behind a guard, you can always r
 <Prompt when={formIsHalfFilledOut} message="Are you sure?"/>
 ```
+
+## beforeUnload: bool _NavigationPrompt_
````
The component is called `<Prompt>` now, not `<NavigationPrompt>`.

mjackson commented on packages/react-router/modules/Prompt.js:
```diff
@@ -39,9 +44,28 @@ class Prompt extends React.Component {
+  enableUnload(_onbeforeunload) {
+    if(this._subscribed) { return }
+    this._subscribed = _onbeforeunload ? _onbeforeunload : this.onbeforeunload;
+    window.addEventListener(BEFOREUNLOAD, this._subscribed)
```
Note that this behavior now makes this component dependent on the DOM, which I'd like to avoid for the generic `<Prompt>` component. Let's do this work in a separate `<BrowserPrompt>` component instead that wraps `<Prompt>`.

gpltaylor replied:
apologies, I assumed that Prompt will aways be used within a Browser. I'm guessing Prompt will also work in Native? 
All the work with Monorepos and Lerna is making sense :)

There are multiple ways to do this, I am going to start working on HOC as I prefer this solution, unless I run into trouble :) Let me know if you think otherwise

- Inheritance
- Composition
- HOC

It seems over the top to create a new Module just for this one feature? Docs, Examples and new code file (even if it is a micro HOC).
You're the boss so I will make a new component :)

mjackson replied:
`<Prompt>` should work everywhere, `<BrowserPrompt>` is a `<Prompt>` that is able to do stuff with the `beforeunload` DOM event. That's how I've thought about it anyway. I'm open to suggestions if you've got different ideas about how it should be done.

mjackson commented on packages/react-router-website/api/Prompt.md:
````diff
@@ -34,3 +34,24 @@ Instead of conditionally rendering a `<Prompt>` behind a guard, you can always r
+
+```js
+<NavigationPrompt beforeUnload={() { 
+  console.log("custom action - i.e. saving data before leaving.")
````
I don't think we want to encourage people to do anything with data in the prompt function, at least not yet.

mjackson commented on packages/react-router-website/api/Prompt.md:
```diff
@@ -34,3 +34,24 @@ Instead of conditionally rendering a `<Prompt>` behind a guard, you can always r
+<NavigationPrompt beforeUnload={() { 
+  console.log("custom action - i.e. saving data before leaving.")
+  var dialogText = "Changes you made may not be saved."
+  e.returnValue = dialogText
```
It would be great if we could provide a higher-level API than `event.returnValue`. We should be able to do that for people if they just return the string they'd like to use in the prompt message, right?

gpltaylor replied:
If you're talking about using the return value, or providing a custom message for the dialog, then this is not possible in the latest browsers. This feature has been dropped. The message here is simply to support older browsers to make them consistent. The return value is overridden in modern browsers.

Starting with Firefox 4, Chrome 51, Opera 38 and Safari 9.1, a generic string not under the control of the webpage will be shown instead of the returned string.

https://developer.mozilla.org/en-US/docs/Web/API/WindowEventHandlers/onbeforeunload

Let me know if I have miss understood your comment.

gpltaylor commented:
Morning @mjackson I am still working on this so not submitted a PR yet.

I have created a HOC that creates the BrowserPrompt component that wraps Prompt.
I have not exposed (exported) the HOC, just the BrowserPrompt created using the HOC.

Sorry for the time taken to create this, I ran into a few issues where SWPrecacheWebpackPlugin would not work on my platform (Window/10) and I could not import the new module (BrowserPrompt) from the website project. Still, I have no idea how this works, I am assuming you have to release the npm to the website somehow?

Works
`import BrowserPrompt from '../../react-router/BrowserPrompt'`
Can't find module BrowserPrompt
`import BrowserPrompt from 'react-router/BrowserPrompt'`

(sorry, I feel like a noobie)

Do you want me to submit another PR so you can review progress, or should I only do this once we are happy with the implementation?

Have a look and let me know if you're happy with this direction.

gpltaylor commented:
The Tests are working on my local build? (work's on my computer :1st_place_medal: :smile:)

mjackson commented:
> I ran into a few issues where SWPrecacheWebpackPlugin would not work on my platform

Sorry about that. I didn't realize that plugin doesn't work with webpack-dev-server. It should be fixed now if you rebase.

> Do you want me to submit another PR so you can review progress, or should I only do this once we are happy with the implementation?

We shouldn't need another PR. As you make changes to your code, just `git push -f` the commits you want to include in the PR to your branch (`gpltaylor/v4`) and this PR will automatically update. We can do all the work here :)

I'll take a closer look here in the morning. Thanks!

gpltaylor commented:
I noticed the latest code has the new project "react-router-dom" package, so I'm guessing the BrowserPromt should live in that project. I will update once I rebase.

mjackson commented:
Thanks, @gpltaylor. I didn't have time to look at this PR today, but will definitely make time to give you some feedback tomorrow.

gpltaylor commented:
hi @mjackson no probs. I know how difficult it can be, didn't get a chance to rebase last night, Will see if I can do this tomorrow. If you can answer the question about the following I would be grateful.

`import BrowserPrompt from '../../react-router/BrowserPrompt'`

mjackson commented:
In `react-router-website` we use paths directly to the `react-router` code so we can build examples for the site that people can just copy/paste directly into create-react-app. What exactly is your question?

gpltaylor commented:
Hi @mjackson, the code fails to find the new module I have added "BrowserPromt" as it does not exist within react-router-dom

However, if I do the following, then it is found, but is technically wrong to do so.

`import BrowserPrompt from '../../react-router-dom/BrowserPrompt'`
This, should be
`import BrowserPrompt from 'react-router-dom/BrowserPrompt'`

However, as I have only just added the file (BrowserPrompt) it does not exist within the react-router-dom npm package.

How is this import getting the current (latest code build) lib/npm and not an NPM downloaded? Is there some configuration I am missing. 
I have done, development on NPM where I've had to release the NPM locally (npm link) but I don't think I should be doing this for react-router?

I have updated the code to the latest, rebase and moved BrowerPrompt to react-router-dom
This all looks good to me, apart from this one issue.

pshrmn commented:
I think that you merged when you wanted to rebase.

gpltaylor commented:
@pshrmn I pulled the latest code down from origin/v4 into a new branch (originv4)
Then I rebase my v4 from originv4
Then I pushed (but got conflicts), I resolved them and pushed, so I think this caused the merge.
Stupidly, I did not go to bed at this point and try again in the morning :)

gpltaylor commented:
Any feedback would be greatly appreciated..

pshrmn commented:
@gpltaylor Not to sound like a broken record, but can you fix the rebase? It is difficult to look through your changes to give feedback when there are 45 commits/73 files changed.

timdorr commented:
This should work (assuming `origin` points to your fork):
```sh
git remote add ReactTraining https://github.com/ReactTraining/react-router
git fetch ReactTraining
git rebase ReactTraining/v4
git push origin v4
```

gpltaylor commented:
Ok, I will see if I can fix this.

gpltaylor commented:
Hopefully this works?

timdorr commented:
Opened up a replacement PR for you that has only the necessary commits.

srirammanohar commented:
Hi is the fix available in the npm package? I dont . see the fix in 4.2.1 or 4.2.2

deeparora29 commented:
Can somebody please let me know in which version of react-router , i will be able to use Browser Prompt??
